File: packages/sdk-python/constitutional/utils/request.py
"""HTTP client for making API requests."""

import time
from typing import Optional, Any, TypeVar
import httpx

from constitutional.types.common import (
    ConstitutionalError,
    RateLimitError,
    AuthenticationError,
    NotFoundError,
)


T = TypeVar("T")

DEFAULT_BASE_URL = "https://api.constitutional.io"
DEFAULT_TIMEOUT = 30.0
DEFAULT_MAX_RETRIES = 3


class HttpClient:
    """HTTP client for Constitutional API."""

    def __init__(
        self,
        api_key: str,
        base_url: Optional[str] = None,
        region: Optional[str] = None,
        timeout: float = DEFAULT_TIMEOUT,
        max_retries: int = DEFAULT_MAX_RETRIES,
    ):
        self.api_key = api_key
        self.base_url = base_url or DEFAULT_BASE_URL
        self.region = region
        self.timeout = timeout
        self.max_retries = max_retries

        if region:
            self.base_url = f"https://{region}.api.constitutional.io"

        self._client = httpx.Client(
            base_url=self.base_url,
            timeout=timeout,
            headers=self._build_headers(),
        )

    def _build_headers(self) -> dict[str, str]:
        """Build request headers."""
        return {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json",
            "X-SDK-Version": "1.0.0",
            "X-SDK-Language": "python",
        }

    def get(
        self,
        path: str,
        params: Optional[dict[str, Any]] = None,
    ) -> dict[str, Any]:
        """Make a GET request."""
        return self._request("GET", path, params=params)

    def post(
        self,
        path: str,
        json: Optional[dict[str, Any]] = None,
    ) -> dict[str, Any]:
        """Make a POST request."""
        return self._request("POST", path, json=json)

    def put(
        self,
        path: str,
        json: Optional[dict[str, Any]] = None,
    ) -> dict[str, Any]:
        """Make a PUT request."""
        return self._request("PUT", path, json=json)

    def delete(self, path: str) -> Optional[dict[str, Any]]:
        """Make a DELETE request."""
        return self._request("DELETE", path)

    def _request(
        self,
        method: str,
        path: str,
        params: Optional[dict[str, Any]] = None,
        json: Optional[dict[str, Any]] = None,
    ) -> dict[str, Any]:
        """Make an HTTP request with retries."""
        url = f"/api{path}"

        # Filter out None values from params
        if params:
            params = {k: v for k, v in params.items() if v is not None}

        last_error: Optional[Exception] = None
        retries = 0

        while retries <= self.max_retries:
            try:
                response = self._client.request(
                    method=method,
                    url=url,
                    params=params,
                    json=json,
                )

                request_id = response.headers.get("x-request-id")

                if response.status_code == 204:
                    return {}

                if response.is_success:
                    return response.json()

                # Handle errors
                try:
                    error_body = response.json()
                except Exception:
                    error_body = {}

                error_message = error_body.get("error", {}).get("message", "Request failed")
                error_code = error_body.get("error", {}).get("code", "ERROR")

                if response.status_code == 429:
                    retry_after = int(response.headers.get("retry-after", "60"))
                    raise RateLimitError(
                        message=error_message,
                        retry_after=retry_after,
                        request_id=request_id,
                    )

                if response.status_code == 401:
                    raise AuthenticationError(
                        message=error_message,
                        request_id=request_id,
                    )

                if response.status_code == 404:
                    raise NotFoundError(
                        resource="Resource",
                        resource_id=path,
                        request_id=request_id,
                    )

                raise ConstitutionalError(
                    message=error_message,
                    code=error_code,
                    status_code=response.status_code,
                    details=error_body.get("error", {}).get("details"),
                    request_id=request_id,
                )

            except AuthenticationError:
                raise

            except RateLimitError as e:
                if retries < self.max_retries:
                    time.sleep(e.retry_after)
                    retries += 1
                    continue
                raise

            except httpx.TimeoutException:
                last_error = ConstitutionalError(
                    message="Request timed out",
                    code="TIMEOUT",
                    status_code=408,
                )
                if retries < self.max_retries:
                    time.sleep(2**retries)
                    retries += 1
                    continue
                raise last_error

            except httpx.HTTPError as e:
                last_error = ConstitutionalError(
                    message=str(e),
                    code="HTTP_ERROR",
                    status_code=500,
                )
                if retries < self.max_retries:
                    time.sleep(2**retries)
                    retries += 1
                    continue
                raise last_error

        if last_error:
            raise last_error
        raise ConstitutionalError("Request failed after retries", "RETRY_EXHAUSTED", 500)

    def close(self) -> None:
        """Close the HTTP client."""
        self._client.close()

    def __enter__(self) -> "HttpClient":
        return self

    def __exit__(self, *args: Any) -> None:
        self.close()

File: packages/sdk-js/src/utils/request.ts
import fetch from 'cross-fetch';
import {
  ConstitutionalConfig,
  ConstitutionalError,
  RateLimitError,
  AuthenticationError,
  NotFoundError,
} from '../types/common.types';

const DEFAULT_BASE_URL = 'https://api.constitutional.io';
const DEFAULT_TIMEOUT = 30000;
const DEFAULT_MAX_RETRIES = 3;

export interface RequestOptions {
  method?: 'GET' | 'POST' | 'PUT' | 'DELETE' | 'PATCH';
  body?: unknown;
  params?: Record<string, string | number | boolean | undefined>;
  headers?: Record<string, string>;
}

export class HttpClient {
  private config: Required<ConstitutionalConfig>;

  constructor(config: ConstitutionalConfig) {
    this.config = {
      apiKey: config.apiKey,
      baseUrl: config.baseUrl || DEFAULT_BASE_URL,
      region: config.region || '',
      timeout: config.timeout || DEFAULT_TIMEOUT,
      maxRetries: config.maxRetries || DEFAULT_MAX_RETRIES,
    };
  }

  async request<T>(path: string, options: RequestOptions = {}): Promise<T> {
    const url = this.buildUrl(path, options.params);
    const headers = this.buildHeaders(options.headers);

    let lastError: Error | null = null;
    let retries = 0;

    while (retries <= this.config.maxRetries) {
      try {
        const controller = new AbortController();
        const timeoutId = setTimeout(() => controller.abort(), this.config.timeout);

        const response = await fetch(url, {
          method: options.method || 'GET',
          headers,
          body: options.body ? JSON.stringify(options.body) : undefined,
          signal: controller.signal,
        });

        clearTimeout(timeoutId);

        const requestId = response.headers.get('x-request-id') || undefined;

        if (response.ok) {
          return await response.json();
        }

        // Handle error responses
        const errorBody = await response.json().catch(() => ({}));

        if (response.status === 429) {
          const retryAfter = parseInt(response.headers.get('retry-after') || '60', 10);
          throw new RateLimitError(
            errorBody.error?.message || 'Rate limit exceeded',
            retryAfter,
            requestId
          );
        }

        if (response.status === 401) {
          throw new AuthenticationError(
            errorBody.error?.message || 'Invalid API key',
            requestId
          );
        }

        if (response.status === 404) {
          throw new NotFoundError('Resource', path, requestId);
        }

        throw new ConstitutionalError(
          errorBody.error?.message || `Request failed with status ${response.status}`,
          errorBody.error?.code || 'REQUEST_FAILED',
          response.status,
          errorBody.error?.details,
          requestId
        );
      } catch (error) {
        lastError = error as Error;

        // Don't retry authentication errors
        if (error instanceof AuthenticationError) {
          throw error;
        }

        // Retry rate limit errors after waiting
        if (error instanceof RateLimitError) {
          if (retries < this.config.maxRetries) {
            await this.sleep(error.retryAfter * 1000);
            retries++;
            continue;
          }
          throw error;
        }

        // Retry network errors
        if (error instanceof Error && error.name === 'AbortError') {
          if (retries < this.config.maxRetries) {
            await this.sleep(Math.pow(2, retries) * 1000);
            retries++;
            continue;
          }
        }

        throw error;
      }
    }

    throw lastError || new Error('Request failed after retries');
  }

  async get<T>(path: string, params?: Record<string, string | number | boolean | undefined>): Promise<T> {
    return this.request<T>(path, { method: 'GET', params });
  }

  async post<T>(path: string, body?: unknown): Promise<T> {
    return this.request<T>(path, { method: 'POST', body });
  }

  async put<T>(path: string, body?: unknown): Promise<T> {
    return this.request<T>(path, { method: 'PUT', body });
  }

  async delete<T>(path: string): Promise<T> {
    return this.request<T>(path, { method: 'DELETE' });
  }

  private buildUrl(path: string, params?: Record<string, string | number | boolean | undefined>): string {
    let baseUrl = this.config.baseUrl;

    // Use regional endpoint if specified
    if (this.config.region) {
      baseUrl = `https://${this.config.region}.api.constitutional.io`;
    }

    const url = new URL(`/api${path}`, baseUrl);

    if (params) {
      Object.entries(params).forEach(([key, value]) => {
        if (value !== undefined) {
          url.searchParams.append(key, String(value));
        }
      });
    }

    return url.toString();
  }

  private buildHeaders(customHeaders?: Record<string, string>): Record<string, string> {
    return {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${this.config.apiKey}`,
      'X-SDK-Version': '1.0.0',
      'X-SDK-Language': 'javascript',
      ...customHeaders,
    };
  }

  private sleep(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }
}
